Document return values of template module

mprasil opened the pull request:
##### SUMMARY

Add documentation for some return values of template module.

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME

template

##### ADDITIONAL INFORMATION
I've only included values that are returned consistently, there are more
that are only returned under special circumstances - like when
destination is changed - but I'm not entirely sure if that behavior is
entirely intentional or subject to change, so I decided to not include
it in this PR.

I have not tested this extensively, so it's possible that under special
circumstances this might not be entirely accurate, but briefly looking
at the code as far as I can tell they are correct.

Sample code I used to see the returned values:

```
$ ansible -m template -a "dest=/tmp/test src=/tmp/template.j2" localhost
localhost | SUCCESS => {
    "changed": false,
    "checksum": "373296322247ab85d26d5d1257772757e7afd172",
    "dest": "/tmp/test",
    "gid": 1000,
    "group": "mprasil",
    "mode": "0664",
    "owner": "mprasil",
    "path": "/tmp/test",
    "size": 22,
    "state": "file",
    "uid": 1000
}

```

Note that I haven't tested this. Not sure how can I generate the docs
locally to test that my change renders correctly.

mprasil committed:
Document return values of template module

Add documentation for some return values of template module. I've only
included values that are returned consistently, there are more that are
only returned under special circumstances - like when destination is
changed - but I'm not entirely sure if that behavior is entirely
intentional or subject to change.

I have not tested this extensively, so it's possible that under special
circumstances this might not be entirely accurate, but briefly looking
at the code as far as I can tell they are correct.

Sample code I used to see the returned values:

```
$ ansible -m template -a "dest=/tmp/test src=/tmp/template.j2" localhost
localhost | SUCCESS => {
    "changed": false,
    "checksum": "373296322247ab85d26d5d1257772757e7afd172",
    "dest": "/tmp/test",
    "gid": 1000,
    "group": "mprasil",
    "mode": "0664",
    "owner": "mprasil",
    "path": "/tmp/test",
    "size": 22,
    "state": "file",
    "uid": 1000
}

```

samccann commented:
/azp run

sivel commented:
I'm on the fence about this change.  The problem is that many of the return values come from a combination of `copy` and `file`.  However, we don't have a great way of indicating that yet.  See https://github.com/ansible/ansible/pull/72635

mprasil commented:
Well I think there are some other modules that also get some of the return values from `copy` and `file` and the return values are documented there? (In fact some of the documentation in this PR was copied from one such module IIRC)

It's possible to incrementally replace the copied stuff with the `RETURN` fragments as (when) they are available. So having the docs already in place would make that easier I imagine?

webknjaz commented:
/azp run

webknjaz commented:
The PR is so old that the CI doesn't fetch enough of the Git tree to reach the common ancestor, causing `git diff` to fail. This is happening @ https://github.com/ansible/ansible/blob/f10d11bcdc54c9b7edc0111eb38c59a88e396d0a/test/lib/ansible_test/_internal/ci/azp.py#L192-L218.

@mprasil could you rebase this branch, so the CI would pick it up?

@mattclay do you think we could implement something like https://stackoverflow.com/a/74710919/595220 or https://github.com/rmacklin/fetch-through-merge-base/blob/98594af/fetch_through_merge_base.sh#L5-L8 in ansible-test?

mattclay commented:
> do you think we could implement something like https://stackoverflow.com/a/74710919/595220 or https://github.com/rmacklin/fetch-through-merge-base/blob/98594af/fetch_through_merge_base.sh#L5-L8 in ansible-test?

This case is rare enough that I don't think we should try to accommodate it. I was going to suggest that we could increase the fetch depth, but this PR is nearly 2000 commits behind, which would be a 4x increase.

mprasil commented:
@webknjaz I have rebased now. It's been a while since I made this PR, I probably lost all the context for the change, but hopefully it's okay as is.

samccann commented:
The docs LGTM but I can't tell if they are accurate. Can someone give a tech review of this?

Commit:
Update the documentation to reflect latest Ansible

THe previous commit was created quite a while ago. It seems that in the
meantime one of the return parameters was renamed from `user` to
`owner`. This should make documentation up to date.

mprasil commented:
@samccann so I tested this with latest version of Ansible 8.2.0 and it turns out one of the return params has been renamed. I added commit 547104b to reflect that.

You can easily test yourself with:

```yaml
# ansible -m template -a "dest=/tmp/test src=/dev/null" localhost
[WARNING]: No inventory was parsed, only implicit localhost is available
localhost | CHANGED => {
    "changed": true,
    "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
    "dest": "/tmp/test",
    "gid": 0,
    "group": "root",
    "md5sum": "d41d8cd98f00b204e9800998ecf8427e",
    "mode": "0644",
    "owner": "root",
    "size": 0,
    "src": "/root/.ansible/tmp/ansible-tmp-1691183563.259344-13-192571672005684/source",
    "state": "file",
    "uid": 0
}
```

It does not need any playbook or extra files, I'm just testing with `/dev/null` as a template for simplicity.The above was executed in docker container with Ansible installed, but your results should be pretty similar.

bcoca commented on lib/ansible/modules/template.py:
```diff
@@ -110,3 +110,46 @@
+    description: Size of the rendered file in bytes
+    returned: success
+    type: int
+    sample: 42
```
missing src (returned when changed), also missing state and md5sum(also on change), though I think we can live w/o those and just deprecate them for this case ...

mprasil replied:
Yeah, I've only included values that are returned consistently.

bcoca replied:
that is what the `returned: success` part is for, which is also incorrect for some entries, should be 'always'

bcoca reviewed (changes requested):
add missing fields, correct the 'returned' entry as some are incorrect

Commit:
Add missing fields and fix issues raised in PR

Some fields were missing and some were incorrectly marked as `returned:
success`. This should fix these issues raised in PR.

mprasil commented:
Thanks for review @bcoca. I have now hopefully corrected all issues. I did the documentation based on this output:

```yaml
# ansible -m template -a "dest=/tmp/test src=/dev/null" localhost
localhost | CHANGED => {
    "changed": true,
    "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
    "dest": "/tmp/test",
    "gid": 0,
    "group": "root",
    "md5sum": "d41d8cd98f00b204e9800998ecf8427e",
    "mode": "0644",
    "owner": "root",
    "size": 0,
    "src": "/root/.ansible/tmp/ansible-tmp-1691603137.1319447-34-199034521062747/source",
    "state": "file",
    "uid": 0
}
# ansible -m template -a "dest=/tmp/test src=/dev/null" localhost
localhost | SUCCESS => {
    "changed": false,
    "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
    "dest": "/tmp/test",
    "gid": 0,
    "group": "root",
    "mode": "0644",
    "owner": "root",
    "path": "/tmp/test",
    "size": 0,
    "state": "file",
    "uid": 0
}
# ansible -m template -a "dest=/tmp/fff/fff/test src=/dev/null" localhost
localhost | FAILED! => {
    "changed": false,
    "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
    "msg": "Destination directory /tmp/fff/fff does not exist"
}
```

bcoca commented on lib/ansible/modules/template.py:
```diff
@@ -110,3 +110,61 @@
+    returned: success
+    type: str
+    sample: 1755
+msg:
```
this is part of the standard, like changed/failed/etc, no need to add here

mprasil replied:
Fixed in [321d9a2](https://github.com/ansible/ansible/pull/75980/commits/321d9a261fe8d8f3cc0bff92038803ff44e2420c).

bcoca commented:
aside from msg, everything else looks correct ... not that i'm happy with the inconsistent returns or src/path overlap .. but that is not an issue the documentation can fix, just reflect. Adding this to my list to eventually fix/normalize outputs.

Commit:
Drop the `msg` field

This is part of the standard, like changed/failed/etc, no need to add
here.

mprasil commented:
@bcoca thanks for review, I've removed the `msg` field now in latest commit.

bcoca commented:
@mprasil thank you for your contribution
